src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -121,11 +124,93 @@ protected override void RemoveItem(int index)
 
             base.RemoveItem(index);
 
+            if (!_skipRaisingEvents)
+            {
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            List<T> removedItems = new List<T>();
+            for (int i = index; i < length; i++)
+            {
+                removedItems.Add(this[i]);
+            }
+
+            bool ignore = _skipRaisingEvents;
+            if (!ignore)
+            {
+                _skipRaisingEvents = true;
+            }
+
+            try
+            {
+                base.RemoveItemsRange(index, count);
+            }
+            finally
+            {
+                if (!ignore)
+                {
+                    _skipRaisingEvents = false;
+                }
+            }
+            
+            if (count > 0 && !_skipRaisingEvents)
+            {
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a collection of items is added to list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void ReplaceItemsRange(int index, int count, IEnumerable<T> collection)
+        {
+            CheckReentrancy();
+
+            _skipRaisingEvents = true;
+
+            List<T> oldCollection = collection is List<T> ?
+                (List<T>)collection :
+                new List<T>(collection);
```

Comment:
The way this logic currently works is the replace event includes 2 IList:
* The Old Collection (which is the snippet here)
* The New Collection after the replacement happens

I was not given explicit business rules on how this should work and would love some input from @ahsonkhan or @safern before we proceed with another set of changes on this.

Reply:
> The way this logic currently works is the replace event includes 2 IList:
> 
> * The Old Collection (which is the snippet here)
> * The New Collection after the replacement happens

I don't think that's right and not how you've currently implemented it. Right now the implementation is passing the same `collection` as both old and new.

I believe the way it's supposed to work is documented here:

https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.notifycollectionchangedeventargs.-ctor?view=netcore-3.0#System_Collections_Specialized_NotifyCollectionChangedEventArgs__ctor_System_Collections_Specialized_NotifyCollectionChangedAction_System_Collections_IList_System_Collections_IList_System_Int32_

- `newItems`: The new items that are replacing the original items.
  - This would be the passed-in `collection` as `IList`.
- `oldItems`: The original items that are replaced.
  - This would be the items removed from the collection.

Reply:
Here is a code snippet of where we pass invoke the vent for ReplaceItemsRange:
```c#
OnCollectionChanged(NotifyCollectionChangedAction.Replace, oldCollection, newCollection, index);
```

Note that it is passing the 2 collections:
* `oldCollection`
* `newCollection`

Thanks for providing links to the docs. I agree that we need to update the implementation to match what the docs state. This is a good catch, I'll update the branch

Reply:
> Note that it is passing the 2 collections:
> 
> * `oldCollection`
> * `newCollection`

Right, but look at how you're initializing `oldCollection` and `newCollection`:

`oldCollection`:
```c#
List<T> oldCollection = collection is List<T> ?
    (List<T>)collection :
    new List<T>(collection);
```

`newCollection`:
```c#
List<T> newCollection = collection is List<T> ?
    (List<T>)collection :
    new List<T>(collection);
```

They're both using the passed-in `collection`.

Reply:
This part of the code is stale, we should really focus our attention on the new code that I am going to be pushing up based on your feedback for NewItems/OldItems instead of the old code.

Reply:
This has been pushed, I'm going to mark this one resolved. If you see issues with the code feel free to reopen